src/acceleration/BaseQNAcceleration.cpp

```diff
@@ -170,7 +174,100 @@ void BaseQNAcceleration::initialize(
 
   _preconditioner->initialize(subVectorSizes);
 }
+void BaseQNAcceleration::projectFWInput(DataMap &cplData, const std::vector<DataID> &dataIDs, std::map<int, double> lowerBounds,
+                                        std::map<int, double> upperBounds)
+{
+  Eigen::Index offset = 0;
+  for (auto id : dataIDs) {
+    Eigen::Index size       = cplData.at(id)->values().size();
+    auto         lowerBound = lowerBounds.at(id);
+    auto         upperBound = upperBounds.at(id);
+    if (lowerBound <= -1.0e16 && upperBound >= 1.0e16) { // do nothing
+    } else if (lowerBound > -1.0e16 && upperBound < 1.0e16) {
+      // when both sides are bounded
+      double delta       = 0.1;
+      double leftLimit   = lowerBound - delta;
+      double rightLimit  = upperBound + delta;
+      double intervalLen = rightLimit - leftLimit;
+      // double factor        = 1;
+      // int selectedTrans = 1; // 1: logTrans, 2: tanhTrans, else: noTrans
+
+      for (Eigen::Index i = 0; i < size; i++) {
+        double normOldValue = (_oldValues[i + offset] - leftLimit) / intervalLen;
+        double normValue    = (_values[i + offset] - leftLimit) / intervalLen;
+
+        _values[i + offset]    = log(normValue / (1.0 - normValue));
+        _oldValues[i + offset] = log(normOldValue / (1.0 - normOldValue));
+        /*
+        switch (selectedTrans) {
+        case 1:
+          _values[i + offset]    = factor * log(normValue / (1.0 - normValue));
+          _oldValues[i + offset] = factor * log(normOldValue / (1.0 - normOldValue));
+          break;
+        case 2:
+          _values[i + offset]    = factor * atanh(normValue);
+          _oldValues[i + offset] = factor * atanh(normOldValue);
+          break;
+        default:
+          _values[i + offset]    = _values[i + offset];
+          _oldValues[i + offset] = _oldValues[i + offset];
+          break;
+        }
+        */
```

Comment:
Always tricky: Which research code goes into the release version, which one not? When you want to compare these for a paper, it is always more helpful to add them and make them configurable. For the user, we can still give a default value to hide the option. Otherwise reproducibility of your published results gets complicated.

Reply:
In the later(complete) version I will remove this things for trying stuff  in the comments ;)